nsxt/resource_nsxt_upgrade_prepare.go

```diff
@@ -0,0 +1,378 @@
+			"upgrade_bundle_url": {
+				Type:        schema.TypeString,
+				Description: "URL of the NSXT Upgrade bundle",
+				Required:    true,
+			},
+			"precheck_bundle_url": {
+				Type:        schema.TypeString,
+				Description: "URL of the NSXT Upgrade precheck bundle (Only applied to NSXT version >= 4.1.1)",
+				Optional:    true,
+			},
+			"version": {
+				Type:        schema.TypeString,
+				Description: "Version for the upgrade",
+				Required:    true,
+			},
+			"accept_user_agreement": {
+				Type:        schema.TypeBool,
+				Description: "Whether to accept the user agreement",
+				Required:    true,
+			},
+			"failed_prechecks": {
+				Type:        schema.TypeList,
+				Description: "List of failed prechecks for the upgrade, only include warnings, if precheck failed with error then the resource creation will fail",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:        schema.TypeString,
+							Description: "ID of failed precheck",
+							Computed:    true,
+						},
+						"message": {
+							Type:        schema.TypeString,
+							Description: "Message of the failed precheck",
+							Computed:    true,
+						},
+					},
+				},
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceNsxtUpgradePrepareCreate(d *schema.ResourceData, m interface{}) error {
+	id := d.Id()
+	if id == "" {
+		id = newUUID()
+	}
+	err := prepareForUpgrade(d, m)
+	if err != nil {
+		return handleCreateError("NsxtUpgradePrepare", id, err)
+	}
+	d.SetId(id)
+	return resourceNsxtUpgradePrepareRead(d, m)
+}
+
+func prepareForUpgrade(d *schema.ResourceData, m interface{}) error {
+	// 1. Upload upgrade bundle and wait for upload to complete
+	err := uploadPrecheckAndUpgradeBundle(d, m)
+	if err != nil {
+		return logAPIError("Failed to upload bundle", err)
+	}
+	// 2. Accept eula
+	err = acceptUserAgreement(d, m)
+	if err != nil {
+		return err
+	}
+	// 3. Upgrade UC and check for its upgrade status
+	err = upgradeUc(d, m)
+	if err != nil {
+		return logAPIError("Failed to upgrade Upgrade Coordinator", err)
+	}
+	// 4. Execute pre-upgrade checks
+	err = executePreupgradeChecks(d, m)
+	if err != nil {
+		return logAPIError("Failed to execute pre-upgrade checks", err)
+	}
+	// 5. Check if there is error in prechecks
+	precheckFailures, err := getPrecheckErrors(m, nsxModel.UpgradeCheckFailure_TYPE_FAILURE)
+	if err != nil {
+		return logAPIError("Failed to get precheck errors", err)
+	}
+	if len(precheckFailures) > 0 {
+		var failureIDs []string
+		for _, failure := range precheckFailures {
+			failureID := *failure.Id
+			failureIDs = append(failureIDs, failureID)
+		}
+		return fmt.Errorf("Encounter errors in precheck %s, please address these error on NSX UI", failureIDs)
+	}
+	return nil
+}
+
+func resourceNsxtUpgradePrepareRead(d *schema.ResourceData, m interface{}) error {
+	id := d.Id()
+	precheckWarnings, err := getPrecheckErrors(m, nsxModel.UpgradeCheckFailure_TYPE_WARNING)
```

Comment:
Are we presenting errors as well somewhere?

Reply:
Also, I wonder if acknowledged prechecks would disappear from the list once acknowledged?

Reply:
If there are errors then the creation of the resource will fail, and id of these errors will be displayed in error message

Reply:
Resolve this conversation as we are displaying both error and warnings in final state of the resource